src/main/java/io/vertx/redis/client/impl/RedisConnectionImpl.java

```diff
@@ -110,19 +111,16 @@ public RedisConnection send(final Request request, Handler<AsyncResult<Response>
 
     // encode the message to a buffer
     final Buffer message = ((RequestImpl) request).encode();
-    final ContextAwareHandler<AsyncResult<Response>> ctxAwareHandler = new ContextAwareHandler<>(handler);
     // all update operations happen inside the context
-    context.runOnContext(v -> {
```

Comment:
Tests are failing and i believe this is the root cause. The queue implementation is not synchonized to make it fast so all mutations must happen on the event loop. Also doing that + socket writes ensures the order is preserved and correct for the callbacks.

A short optimization here is to verify that were on the same context as the socket and avoid the run on context.

Reply:
Why do we need synchronization?

As explained above all methods should be called from the same event loop.

Please correct me if i am wrong

Reply:
Could you please tell me which tests are failing for you?

Reply:
https://travis-ci.org/vert-x3/vertx-redis-client/builds/612915093

Reply:
onConnect callback that creates the netClient was called without a context.

That resulted to each netClient instance being executed on a different context and hence different thread.

In the case of clustered tests, 6 connection where initialised on 6 different threads

In order to fix the issue i either had to:
- synchronise the queue (Either by synchronised or runOnContext)
- Run the netClient on the same context as the the RedisClient.create(vertx)

I chose the later, but with some very simple benchmarks on the intensive test cases (RedisClusterTest.runTheSlotScope) this implementation takes on average 200ms more to complete

I also checked vertx-sql-client and there the netClient is also created on the same context as the PgClient.create(vertx)

Reply:
The tests failing on Sentinel is coincidental. The first commit that fixed the problem passed the tests just fine but the second commit witch only added some tests failed on sentinelTest (A test that was not changed between the first and the second commit)